common/cuda_hip/matrix/bccoo_helper.hpp.inc

```diff
@@ -0,0 +1,263 @@
+inline __device__ void loop_block_multi_row(
+    const uint8* __restrict__ chunk_data, size_type block_size_local,
+    const ValueType* __restrict__ b, const size_type b_stride,
+    const size_type column_id, ValueType* __restrict__ c,
+    const size_type c_stride, compr_idxs& idxs, compr_blk_idxs& blk_idxs,
+    const size_type start_in_blk, const size_type jump_in_blk, Closure scale)
+{
+    auto next_row = blk_idxs.row_frs;
+    auto last_row = blk_idxs.row_frs;
+    ValueType temp_val = zero<ValueType>();
+    ValueType val;
+    bool new_value = false;
+    const auto tile_block =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
+
+    last_row = blk_idxs.row_frs +
+               get_value_chunk<IndexType1>(
+                   chunk_data, blk_idxs.shf_row +
+                                   (block_size_local - 1) * sizeof(IndexType1));
+    next_row =
+        blk_idxs.row_frs +
+        get_value_chunk<IndexType1>(
+            chunk_data, blk_idxs.shf_row + start_in_blk * sizeof(IndexType1));
+    for (size_type pos = start_in_blk; pos < block_size_local;
+         pos += jump_in_blk) {
+        idxs.row = next_row;
+        idxs.col = blk_idxs.col_frs +
+                   get_value_chunk<IndexType2>(
+                       chunk_data, blk_idxs.shf_col + pos * sizeof(IndexType2));
+        val = get_value_chunk<ValueType>(
+            chunk_data, blk_idxs.shf_val + pos * sizeof(ValueType));
+        temp_val += val * b[idxs.col * b_stride + column_id];
+        new_value = true;
+        next_row =
+            ((pos + jump_in_blk) >= block_size_local)
+                ? last_row
+                : (blk_idxs.row_frs +
+                   get_value_chunk<IndexType1>(
+                       chunk_data, blk_idxs.shf_row + (pos + jump_in_blk) *
+                                                          sizeof(IndexType1)));
+        // segmented scan
+        if (tile_block.any(idxs.row != next_row)) {
+            bool is_first_in_segment = segment_scan<subwarp_size>(
+                tile_block, idxs.row, temp_val,
+                [](ValueType a, ValueType b) { return a + b; });
+            if (is_first_in_segment) {
+                atomic_add(&(c[idxs.row * c_stride + column_id]),
+                           scale(temp_val));
+            }
+            temp_val = zero<ValueType>();
+            new_value = false;
+        }
+    }
+    // segmented scan
+    if (tile_block.any(new_value)) {
+        bool is_first_in_segment = segment_scan<subwarp_size>(
+            tile_block, idxs.row, temp_val,
+            [](ValueType a, ValueType b) { return a + b; });
+        if (is_first_in_segment) {
+            atomic_add(&(c[idxs.row * c_stride + column_id]), scale(temp_val));
+        }
+        temp_val = zero<ValueType>();
+    }
+}
+
+
+template <typename IndexType, typename ValueType>
+inline GKO_ATTRIBUTES void get_block_position_value(
```

Comment:
Only GPUs should use these methods. therefore, I have used __device__ instead of GKO_ATTRIBUTES
I made the same change in common/cuda_hip/base/unaligned_access.hpp.inc

Reply:
Please, resolve if you agree